### ISSUE TITLE
Support generic protocol access log formatter extension

### ISSUE THREAD

**Issue by caitong93** [envoyproxy/envoy#13085]
*Title*: *Support generic protocol access log formatter extension*

*Description*:

I want to add access log support for dubbo and thrift filter, but find out current access log formatter only support http,  and it cannot be extended. Any thoughts about implement a generic access log formatter ?

**Comment by mattklein123** [comment]
See https://github.com/envoyproxy/envoy/blob/308619fc7a2c403ff674e2af7dda4e73b00575ed/api/envoy/config/accesslog/v3/accesslog.proto#L318

**Comment by caitong93** [comment]
`ExtensionFilter` seems not what I want. `Envoy::AccessLog::Instance`, `Envoy::AccessLog::Formatter`, `Envoy::AccessLog::FormatterProvider`, and implementations of these interfaces,  only support HTTP.  I'm not sure if envoy have plan to refactor these interfaces, so that other protocols can reuse them.

My current plan for implement a thrift access logger, without modifying envoy code:
1. implement a  access logger extension named `envoy.thrift_file_access_log`
2. convert thrift metadata to Http::HeaderMap to use the method `void log(const Http::HeaderMap* request_headers, const Http::HeaderMap* response_headers,
                   const Http::HeaderMap* response_trailers,
                   const StreamInfo::StreamInfo& stream_info)` 

will try to do this as an internal implementation. If anyone is interested, I can send out a pr.

**Comment by lizan** [comment]
This seems a valid feature request but need design proposal. @mattklein123 the extension filter is about how we filter log entries, but it doesn't let access log support generic protocol, other than TCP / HTTP.

@cpakulski was looking into how to integrate MySQL logs (e.g. transaction logs which is in different layers than TCP), we saw same issue with current logging structure including ALS. It makes sense to have some extensible way to let TCP filters to emit custom logs easily (to access logger extensions, file/gRPC/etc).

**Comment by mattklein123** [comment]
> It makes sense to have some extensible way to let TCP filters to emit custom logs easily (to access logger extensions, file/gRPC/etc).

I agree. We actually need access logs for the DNS filter also at Lyft (cc @abaptiste) so whatever we come up with here can be used for that.

Naively I would think that the general HTTP/TCP system should work with possible protocol specific data extensions?

**Comment by cpakulski** [comment]
I am working on a prototype. The idea is to emit fairly flexible messages describing non-HTTP transactions, like database interactions. I could not fit them into gRPC message, as it is very HTTP specific.

**Comment by caitong93** [comment]
add ref to https://github.com/envoyproxy/envoy/pull/14175 , which might be related to this issue

**Comment by wbpcode** [comment]
Is there any progress here ?  GRPC does make this problem very complex. Perhaps we could start by creating a generic file access logger for none HTTP protocol and gradually extend it to other access loggers such as GRPC.

@cpakulski @lizan @mattklein123
